crypto/s2n_rsa_pss.c

```diff
@@ -193,7 +199,7 @@ int s2n_evp_pkey_to_rsa_pss_public_key(struct s2n_rsa_key *rsa_key, EVP_PKEY *pk
 
 int s2n_evp_pkey_to_rsa_pss_private_key(struct s2n_rsa_key *rsa_key, EVP_PKEY *pkey)
 {
-    RSA *priv_rsa_key = EVP_PKEY_get0_RSA(pkey);
+    RSA *priv_rsa_key = EVP_PKEY_get1_RSA(pkey);
     POSIX_ENSURE_REF(priv_rsa_key);
```

Comment:
Are you sure we don't want RSA* to be const? Now that the behavior has changed and modifying the RSA key isn't reflected back in the EVP_PKEY, isn't this likely to cause bugs? How do you know we're not relying on that behavior?

Reply:
This usage is consistent with [rsa](https://github.com/aws/s2n-tls/blob/main/crypto/s2n_rsa.c#L171) and [ecdsa](https://github.com/aws/s2n-tls/blob/main/crypto/s2n_ecdsa.c#L162). That means that we make similar assumptions in all the other implementations also:

The assumptions in ossl1:
- we are getting a raw ref to the key
- the ref count is incremented so we need to call free
- since its a raw ref, we can mutate it

The assumptions in ossl3:
- we are getting a ref to a pre-cached copy of the key
- the ref count of the copy is incremented so we need to call free
- we can can mutate the copy, but the changes are not reflected back to the original key
  - this is ok since we dont modify the key (except free) (based on my investigation of methods allowed on keys)
  - this is ok **if** we store the copy in s2n_rsa_key(the copy) and dont actually use the original

---
In conclusion, why PR impl is Safe but brittle:
- get1 allows us to get around the api changes introduced with get0
  - this forces us to call free on the key
- ossl3 now returns a pre-cached copy, but that is ok since the key is not mutated
- if we do mutate then we violate the above behavior
---
Alternative and reasons why I didnt go that route:
- use PKEY directly: this is a large refactor change **(if we are really uncomfortable with the current behavior then this would be the recommended approach. This can/should also be follow up work since the above is not a 1-way door)**
- use get0 and const: this is a large refactor change since `s2n_rsa.c` and `rsa_signing` will need to change. its also less prefered to using PKEY directly imo

Reply:
As discussed offline, "if we do mutate then we violate the above behavior" is what I'm concerned about. The new Openssl 3.0 behavior effectively makes the RSA* object const, but that's not reflected anywhere in our code. Nothing will stop a future dev from trying to modify the key, and things will not behave as someone reading this code would expect them to. This seems like a recipe for bugs.

If we're not going to make RSA* const, then we minimally need very scary comments everywhere that they key is referenced.

Reply:
specifically here is an impl which stores PKEY and converts to RSA when needed. Its also the recommended ossl approach: https://github.com/aws/s2n-tls/commit/b9fd783b135f3db0c5e40f565d3812ec9758a8c0